Encryption/SAFER-MO/safer.1
.TH SAFER 1 "Sep. 9, 1995"
.GN 2
.SH NAME
safer \- encryption and decryption using SAFER
.SH SYNOPSIS
.TP
.B safer
[
.B \-e
|
.B \-d
] [
.B \-ecb
|
.B \-cbc
|
.B \-cfb
|
.B \-ofb
]
.br
(
.B \-k
.I keyString
|
.B \-kx
.I keyHexString
) 
.br
[
.B \-r
.I nofRounds
] [ 
.B \-s
] [ 
.B \-v
]
.br
[
.I inputFile
[
.I outputFile
] ]
.TP
.B safer
[
.B \-h
|
.B \-hc
] [
.B \-tan
|
.B \-abr
]
.br
[
.B \-k
.I keyString
|
.B \-kx
.I keyHexString
]
.br
[
.B \-r
.I nofRounds
] [
.B \-s
] [
.B \-v
]
.br
[
.I inputFile
[ [
.I outputFile
]
.I hashvalFile
]
.SH DESCRIPTION
.I Safer
reads
.I inputFile
and writes the encrypted or decrypted data to
.I outputFile
or writes the hash value to
.I hashvalFile\c
\&.
If a file name is not given in command line, 
.I safer 
uses the standard input or output.
If the input file name is given as '-', 
.I safer 
reads from the standard input.
.br

.br
.B SAFER
[1] (Secure And Fast Encryption Routine) is a block cipher developed 
by Prof. J.L. Massey at the Swiss Federal Institute of Technology.
There exist four versions of this algorithm, namely: 
.B SAFER K-64
[1], 
.B SAFER K-128
[2],
.B SAFER SK-64
[3] and
.B SAFER SK-128
[3]. The numerals 64 and 128 stand for the length of the user-selected
key, 'K' stands for the original key schedule and 'SK' stands for the
strengthened key schedule (in which some of the "weaknesses" of the
original key schedule have been removed).
.SH OPTIONS
.TP
.B \-e
Encryption (default).
.TP
.B \-d 
Decryption.
.TP
.B \-k 
The key is specified with 
.I keyString\c
\&.
If the length of
.I keyString
is less than 10 characters,
.I keyString
is interpreted as a 64-bit key, otherwise as a 128-bit key. 
.TP
.B \-kx 
The key is specified with
.I keyHexString\c
\&.
If the length of
.I keyHexString
is less than 17 hex digits,
.I keyHexString
is interpreted as a 64-bit key, otherwise as a 128-bit key. 
For the modes 
.B -cbc\c
, 
.B -cfb
and
.B -ofb\c
, it is possible to specify an initial value denoted by y[0]. In this 
case the key and the initial value are separated by a colon, e.g. '1234:9A'.
.TP
.B \-r
.I nofRounds
gives the number of rounds in the encryption (resp. decryption) process.
Default values are 6 rounds for 
.B SAFER  K-64\c
, 8 rounds for
.B SAFER SK-64 
and 10 rounds for
.B SAFER K-128 and
.B SAFER SK-128\c
\&. 
.TP
.B -s
The strengthened key schedule contained in
.B SAFER SK-64
or
.B SAFER SK-128
is used,
instead of the original key schedule contained in
.B SAFER K-64
or
.B SAFER K-128\c
\&.
.TP
.B \-h 
Compute a 128-bit hash value from the input data. The
hash value is written to 
.I hashvalFile
(or standard output).
.TP
.B \-hc 
Compute a 128-bit hash value from the input data. The input is copied to 
.I outputFile
(or standard output) and the hash value is written to 
.I hashvalFile
(or standard error).
.TP
.B \-v
Verbose mode. The selected parameters are written to standard error.
.TP
.B Notation:
.br
.ta 1.3i 1.5i
.in +1.5i
.ti -1.5i
z	=	64-bit or 128-bit key
.br
.ti -1.5i
x[i]	=	i-th 64-bit plaintext block (i = 1..L+1)
.br
.ti -1.5i
y[i]	=	i-th 64-bit ciphertext block (i = 1..L+1)
.br
.ti -1.5i
x[1]..x[L]	=	actual plaintext (last block is filled with zeros)
.br
.ti -1.5i
x[L+1]	=	length of actual plaintext in bits
.br
.ti -1.5i
x[1]..x[L+1]	=	plaintext
.br
.ti -1.5i
y[1]..y[L+1]	=	ciphertext
.br
.ti -1.5i
<a, b>	=	128-bit block composed of two 64-bit blocks
.br
.ti -1.5i
E(z, .)	=	encryption function under the key z
.br
.ti -1.5i
D(z, .)	=	decryption function, x = D(z, E(z, x))
.br
.ti -1.5i
^	=	bit-by-bit exclisive-OR
.br
.ti -1.5i
~	=	bit-by-bit complement
.in -1.5i
.TP
.B Encryption / Decryption Modes:
.TP
.B \-ecb 
electronic code book mode
.br
y[i] = E(z, x[i])
.br
x[i] = D(z, y[i])
.TP
.B \-cbc 
cipher block chaining mode (default)
.br
y[i] = E(z, x[i] ^ y[i-1])
.br
x[i] = D(z, y[i]) ^ y[i-1]
.TP
.B \-cfb 
ciphertext feedback mode
.br
y[i] = x[i] ^ E(z, y[i-1])
.br
x[i] = y[i] ^ E(z, y[i-1])
.TP
.B \-ofb 
output feedback mode
.br
h[i] = E(z, h[i-1])
.br
y[i] = x[i] ^ h[i]
.br
x[i] = y[i] ^ h[i]
.TP
.B Hash Functions:
.br
If no key is given, 
.I safer 
uses the all zero key.
.br
<h[0], g[0]> = z
.br
hash value = <h[L+1], g[L+1]>
.TP
.B \-tan 
tandem Davies-Meyer scheme (default)
.br
w[i] = E(<g[i-1], x[i]>, h[i-1])
.br
h[i] = h[i-1] ^ w[i]
.br
g[i] = g[i-1] ^ E(<x[i], w[i]>, g[i-1])
.TP
.B \-abr 
abreast Davies-Meyer scheme
.br
h[i] = h[i-1] ^ E(<g[i-1], x[i]>, h[i-1])
.br
g[i] = g[i-1] ^ E(<x[i], h[i-1]>, ~g[i-1])
.TP
.B Key Formats:
.br
.I keyHexString
= z:y[0] = { '0'..'9' | 'a'..'f' | 'A'..'F' | ':' } 
.br
.I keyString
= z = { ' '..'~' }
.SH EXAMPLES
To encrypt and decrypt a file in ciphertext feedback mode:
.LP
.RS
.nf
.ft B
safer -e -cfb -kx 123456:cDd7 data data.cr
safer -d -cfb -kx 123456:cDd7 data.cr data.ori
data and data.ori are identical
.ft R
.fi
.RE
.br
.ne 10
.LP
To compute the hash value:
.LP
.RS
.nf
.ft B
safer -h data
.ft R
.fi
.RE
.br
.ne 10
.LP
To compute the hash value and encrypt the data in one step:
.LP
.RS
.nf
.ft B
safer -hck "k e y" data | safer -kx 12E3 - data.cr
.ft R
.fi
.RE
.SH PATENT
"Although our design of
.B SAFER K-64
was sponsored by Cylink Corporation (Sunnyvale, CA, USA), Cylink has
explicitly relinquished any proprietary rights to this algorithm.
This largesse on the part of Cylink was motivated by the reasoning that
the company would gain more from new business than it would lose from
competition should many new users adopt this publicly available cipher.
.B SAFER K-64
has not been patented and, to the best of our knowledge, is free for use
by anyone without fees of any kind and with no violation of any rights of
ownership, intellectual or otherwise." [2]
.SH REFERENCES
.TP
.B [1]
Massey, J.L.,
"SAFER K-64: A Byte-Oriented Block Ciphering Algorithm",
pp. 1-17 in
.I Fast Software Encryption
(Ed. R. Anderson), Proceedings of the Cambridge Security Workshop,
Cambridge, U.K., Dec. 9-11, 1993, Lecture Notes in Computer Science No. 809.
Heidelberg and New York: Springer, 1994.
.br
.TP
.B [2]
Massey, J.L.,
"SAFER K-64: One Year Later",
preliminary manuscript of a paper presented at the K. U. Leuven Workshop
on Cryptographic Algorithms, Dec. 14-16, 1994. To be published in the
Proceedings of this workshop by Springer.
.br
.TP
.B [3]
Massey, J.L.,
"Announcement of a Strengthened Key Schedule for the Cipher SAFER",
Sept. 9, 1995,
(see file 'SAFER_SK.TXT' on distribution).
.SH AUTHOR
Richard De Moliner (demoliner@isi.ee.ethz.ch)
.br
Signal and Information Processing Laboratory
.br
Swiss Federal Institute of Technology
.br
CH-8092 Zurich, Switzerland
.SH BUGS
This program is at most as strong as SAFER itself. So, we urge the user to 
use this program only after he has assured himself that SAFER is strong 
enough AND he has read the source code of this program and its libraries AND 
he has compiled the program himself with a troyan-free compiler. WE DO NOT 
GUARANTEE THAT THIS PROGRAM IS A SECURE ENCRYPTION PROGRAM.


Encryption/crypto32/safer.cpp
// safer.cpp - modified by by Wei Dai from Richard De Moliner's safer.c

#include "pch.h"
#include "safer.h"

NAMESPACE_BEGIN(CryptoPP)

const byte SAFER::exp_tab[256] = 
	{1, 45, 226, 147, 190, 69, 21, 174, 120, 3, 135, 164, 184, 56, 207, 63,
	8, 103, 9, 148, 235, 38, 168, 107, 189, 24, 52, 27, 187, 191, 114, 247,
	64, 53, 72, 156, 81, 47, 59, 85, 227, 192, 159, 216, 211, 243, 141, 177,
	255, 167, 62, 220, 134, 119, 215, 166, 17, 251, 244, 186, 146, 145, 100, 131,
	241, 51, 239, 218, 44, 181, 178, 43, 136, 209, 153, 203, 140, 132, 29, 20,
	129, 151, 113, 202, 95, 163, 139, 87, 60, 130, 196, 82, 92, 28, 232, 160,
	4, 180, 133, 74, 246, 19, 84, 182, 223, 12, 26, 142, 222, 224, 57, 252,
	32, 155, 36, 78, 169, 152, 158, 171, 242, 96, 208, 108, 234, 250, 199, 217,
	0, 212, 31, 110, 67, 188, 236, 83, 137, 254, 122, 93, 73, 201, 50, 194,
	249, 154, 248, 109, 22, 219, 89, 150, 68, 233, 205, 230, 70, 66, 143, 10,
	193, 204, 185, 101, 176, 210, 198, 172, 30, 65, 98, 41, 46, 14, 116, 80,
	2, 90, 195, 37, 123, 138, 42, 91, 240, 6, 13, 71, 111, 112, 157, 126,
	16, 206, 18, 39, 213, 76, 79, 214, 121, 48, 104, 54, 117, 125, 228, 237,
	128, 106, 144, 55, 162, 94, 118, 170, 197, 127, 61, 175, 165, 229, 25, 97,
	253, 77, 124, 183, 11, 238, 173, 75, 34, 245, 231, 115, 35, 33, 200, 5,
	225, 102, 221, 179, 88, 105, 99, 86, 15, 161, 49, 149, 23, 7, 58, 40};

const byte SAFER::log_tab[256] = 
	{128, 0, 176, 9, 96, 239, 185, 253, 16, 18, 159, 228, 105, 186, 173, 248,
	192, 56, 194, 101, 79, 6, 148, 252, 25, 222, 106, 27, 93, 78, 168, 130,
	112, 237, 232, 236, 114, 179, 21, 195, 255, 171, 182, 71, 68, 1, 172, 37,
	201, 250, 142, 65, 26, 33, 203, 211, 13, 110, 254, 38, 88, 218, 50, 15,
	32, 169, 157, 132, 152, 5, 156, 187, 34, 140, 99, 231, 197, 225, 115, 198,
	175, 36, 91, 135, 102, 39, 247, 87, 244, 150, 177, 183, 92, 139, 213, 84,
	121, 223, 170, 246, 62, 163, 241, 17, 202, 245, 209, 23, 123, 147, 131, 188,
	189, 82, 30, 235, 174, 204, 214, 53, 8, 200, 138, 180, 226, 205, 191, 217,
	208, 80, 89, 63, 77, 98, 52, 10, 72, 136, 181, 86, 76, 46, 107, 158,
	210, 61, 60, 3, 19, 251, 151, 81, 117, 74, 145, 113, 35, 190, 118, 42,
	95, 249, 212, 85, 11, 220, 55, 49, 22, 116, 215, 119, 167, 230, 7, 219,
	164, 47, 70, 243, 97, 69, 103, 227, 12, 162, 59, 28, 133, 24, 4, 29,
	41, 160, 143, 178, 90, 216, 166, 126, 238, 141, 83, 75, 161, 154, 193, 14,
	122, 73, 165, 44, 129, 196, 199, 54, 43, 127, 67, 149, 51, 242, 108, 104,
	109, 240, 2, 40, 206, 221, 155, 234, 94, 153, 124, 20, 134, 207, 229, 66,
	184, 64, 120, 45, 58, 233, 100, 31, 146, 144, 125, 57, 111, 224, 137, 48};

#define EXP(x)       exp_tab[(x)]
#define LOG(x)       log_tab[(x)]
#define PHT(x, y)    { y += x; x += y; }
#define IPHT(x, y)   { x -= y; y -= x; }

SAFER::SAFER(const byte *userkey_1, const byte *userkey_2, unsigned nof_rounds, bool strengthened)
	: keySchedule(1 + BLOCKSIZE * (1 + 2 * nof_rounds))
{
	unsigned int i, j;
	byte *key = keySchedule;
	SecByteBlock ka(BLOCKSIZE + 1), kb(BLOCKSIZE + 1);

	if (MAX_ROUNDS < nof_rounds)
		nof_rounds = MAX_ROUNDS;
	*key++ = (unsigned char)nof_rounds;
	ka[BLOCKSIZE] = 0;
	kb[BLOCKSIZE] = 0;
	for (j = 0; j < BLOCKSIZE; j++)
	{
		ka[BLOCKSIZE] ^= ka[j] = rotlFixed(userkey_1[j], 5U);
		kb[BLOCKSIZE] ^= kb[j] = *key++ = userkey_2[j];
	}
	for (i = 1; i <= nof_rounds; i++)
	{
		for (j = 0; j < BLOCKSIZE + 1; j++)
		{
			ka[j] = rotlFixed(ka[j], 6U);
			kb[j] = rotlFixed(kb[j], 6U);
		}
		for (j = 0; j < BLOCKSIZE; j++)
			if (strengthened)
				*key++ = (ka[(j + 2 * i - 1) % (BLOCKSIZE + 1)]
								+ exp_tab[exp_tab[18 * i + j + 1]]) & 0xFF;
			else
				*key++ = (ka[j] + exp_tab[exp_tab[18 * i + j + 1]]) & 0xFF;
		for (j = 0; j < BLOCKSIZE; j++)
			if (strengthened)
				*key++ = (kb[(j + 2 * i) % (BLOCKSIZE + 1)]
								+ exp_tab[exp_tab[18 * i + j + 10]]) & 0xFF;
			else
				*key++ = (kb[j] + exp_tab[exp_tab[18 * i + j + 10]]) & 0xFF;
	}
}

void SAFER::Encrypt(const byte *block_in, byte *block_out) const
{
	byte a, b, c, d, e, f, g, h, t;
	const byte *key = keySchedule+1;
	unsigned int round = keySchedule[0];

	a = block_in[0]; b = block_in[1]; c = block_in[2]; d = block_in[3];
	e = block_in[4]; f = block_in[5]; g = block_in[6]; h = block_in[7];
	while(round--)
	{
		a ^= key[0]; b += key[1]; c += key[2]; d ^= key[3];
		e ^= key[4]; f += key[5]; g += key[6]; h ^= key[7];
		a = EXP(a) + key[ 8]; b = LOG(b) ^ key[ 9];
		c = LOG(c) ^ key[10]; d = EXP(d) + key[11];
		e = EXP(e) + key[12]; f = LOG(f) ^ key[13];
		g = LOG(g) ^ key[14]; h = EXP(h) + key[15];
		key += 16;
		PHT(a, b); PHT(c, d); PHT(e, f); PHT(g, h);
		PHT(a, c); PHT(e, g); PHT(b, d); PHT(f, h);
		PHT(a, e); PHT(b, f); PHT(c, g); PHT(d, h);
		t = b; b = e; e = c; c = t; t = d; d = f; f = g; g = t;
	}
	a ^= key[0]; b += key[1]; c += key[2]; d ^= key[3];
	e ^= key[4]; f += key[5]; g += key[6]; h ^= key[7];
	block_out[0] = a & 0xFF; block_out[1] = b & 0xFF;
	block_out[2] = c & 0xFF; block_out[3] = d & 0xFF;
	block_out[4] = e & 0xFF; block_out[5] = f & 0xFF;
	block_out[6] = g & 0xFF; block_out[7] = h & 0xFF;
}

void SAFER::Decrypt(const byte *block_in, byte *block_out) const
{
	byte a, b, c, d, e, f, g, h, t;
	unsigned int round = keySchedule[0];
	const byte *key = keySchedule + BLOCKSIZE * (1 + 2 * round) - 7;

	a = block_in[0]; b = block_in[1]; c = block_in[2]; d = block_in[3];
	e = block_in[4]; f = block_in[5]; g = block_in[6]; h = block_in[7];
	h ^= key[7]; g -= key[6]; f -= key[5]; e ^= key[4];
	d ^= key[3]; c -= key[2]; b -= key[1]; a ^= key[0];
	while (round--)
	{
		key -= 16;
		t = e; e = b; b = c; c = t; t = f; f = d; d = g; g = t;
		IPHT(a, e); IPHT(b, f); IPHT(c, g); IPHT(d, h);
		IPHT(a, c); IPHT(e, g); IPHT(b, d); IPHT(f, h);
		IPHT(a, b); IPHT(c, d); IPHT(e, f); IPHT(g, h);
		h -= key[15]; g ^= key[14]; f ^= key[13]; e -= key[12];
		d -= key[11]; c ^= key[10]; b ^= key[9]; a -= key[8];
		h = LOG(h) ^ key[7]; g = EXP(g) - key[6];
		f = EXP(f) - key[5]; e = LOG(e) ^ key[4];
		d = LOG(d) ^ key[3]; c = EXP(c) - key[2];
		b = EXP(b) - key[1]; a = LOG(a) ^ key[0];
	}
	block_out[0] = a & 0xFF; block_out[1] = b & 0xFF;
	block_out[2] = c & 0xFF; block_out[3] = d & 0xFF;
	block_out[4] = e & 0xFF; block_out[5] = f & 0xFF;
	block_out[6] = g & 0xFF; block_out[7] = h & 0xFF;
}

NAMESPACE_END
